ticketbuyer/purchase.go

```diff
@@ -42,27 +45,66 @@ func (p *PurchaseManager) purchase(height int64) {
 
 // NotificationHandler handles notifications, which trigger ticket purchases.
 func (p *PurchaseManager) NotificationHandler() {
+	p.quitMtx.Lock()
+	quit := p.quit
+	p.quitMtx.Unlock()
+
 	s1, s2 := make(chan struct{}), make(chan struct{})
 	close(s1) // unblock first worker
 out:
 	for {
 		select {
 		case v := <-p.ntfnChan:
+			p.wg.Add(1)
 			go func(s1, s2 chan struct{}) {
-				<-s1 // wait for previous worker to finish
+				defer p.wg.Done()
+				select {
+				case <-s1: // wait for previous worker to finish
 
-				// Purchase tickets for each attached block, not just for the
-				// update to the main chain.  This is probably not optimal but
-				// it matches how dcrticketbuyer worked.
-				for h := v.NewHeight - int32(len(v.AttachedBlocks)) + 1; h <= v.NewHeight; h++ {
-					p.purchase(int64(h))
+					// Purchase tickets for each attached block, not just for the
+					// update to the main chain.  This is probably not optimal but
+					// it matches how dcrticketbuyer worked.
+					for h := v.NewHeight - int32(len(v.AttachedBlocks)) + 1; h <= v.NewHeight; h++ {
+						p.purchase(int64(h))
+					}
+				case <-quit:
+					return
 				}
 
 				close(s2) // unblock next worker
 			}(s1, s2)
 			s1, s2 = s2, make(chan struct{})
-		case <-p.quit:
+		case <-quit:
 			break out
 		}
 	}
+	p.wg.Done()
+}
+
+// Start starts the purchase manager goroutines.
+func (p *PurchaseManager) Start() {
+	p.wg.Add(1)
+	go p.NotificationHandler()
+
+	log.Infof("Starting ticket buyer")
+}
+
+// WaitForShutdown blocks until all purchase manager goroutines have finished executing.
+func (p *PurchaseManager) WaitForShutdown() {
+	p.wg.Wait()
+}
+
+// Stop signals all purchase manager goroutines to shutdown.
+func (p *PurchaseManager) Stop() {
+	p.quitMtx.Lock()
+	quit := p.quit
+	p.quitMtx.Unlock()
```

Comment:
this entire function needs to be protected by the mutex. having two goroutines enter the select below is incorrect